### ISSUE TITLE
Upgrade WETH contract

### ISSUE THREAD

**Issue by abandeali1** [0xProject/ZEIPs#14]
## Summary
The current [WETH contract](https://github.com/0xProject/contracts/blob/master/contracts/tokens/EtherToken.sol) supported by the community does not emit events on ETH deposit or withdrawal. These actions involve minting new WETH tokens, or burning them (in return for the equivalent amount of Ether). Many dApps adjust a users balance when receiving ERC20 standard events “Transfer” and “Approval”. If the user deposits or withdraws WETH **on a different dApp**, there is no way other then polling the users WETH balance to react to this event. In order to enable dApps to react to WETH deposits and withdrawals without polling, events should be added to these two methods notifying interested parties of their occurrence.

In addition, an ERC20 specification was finalized after the deployment of our current WETH contract in [EIP20](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-20-token-standard.md#transfer). Our current WETH contract currently returns false on invalid transfers, whereas EIP20 says that a token should throw on invalid transfers. WETH should be updated to stay current with the official specification.

### Reacting to WETH changes occurring off-platform
As more and more dApps standardize around using the same WETH contract, the likelihood that a user deposits/withdraws WETH on multiple services increases. If the user has their browser window open on two dApps and performs an action on their WETH, the only dApps will not be able to react the their altered balance without polling for it on an interval. This will require the user to refresh their page in order to see their altered WETH balance.

### Requisite for building efficient order watchers
In order to react to the changing validity of existing orders, one must efficiently watch for any of the orders dependent blockchain state to change. For 0x protocol orders, this involves reacting to the order maker's balance of ZRX and the makerToken, as well as order fills and cancels. When makerToken's are standard ERC20 tokens, we can simply listen for the “Transfer” and “Approval” events to know when an orders validity might have changed. If the makerToken is WETH however, in addition to these events, an order might become invalid due to a withdrawal of WETH back to ETH. In order to handle this efficiently, it would be most helpful if an event were emitted by the WETH contract. 

## Specification
EIP20 specifies which event should be logged when new tokens are minted: “A token contract which creates new tokens SHOULD trigger a `Transfer` event with the `_from` address set to `0x0` when tokens are created.”

Therefore, the `Transfer` event will be emitted when new tokens are minted or burned. The `_from` address will be set to `0x0` when WETH tokens are minted, and the `_to` address will be set to `0x0` when WETH tokens are burned.

With this information, dApps can adjust a users WETH balance when ETH is deposited or withdrawn from the WETH contract. In addition, an orderWatcher can easily decide if an order involving WETH is no longer valid.

The `transfer` function will now throw if the `_from` account balance does not have enough tokens to spend. `transferFrom` will also throw if the `_from` account balance does not have enough tokens to spend or if the `_from` account did not approve the sender of the message.

**Comment by MicahZoltu** [comment]
I argued against the __SHOULD__ in the ERC20 spec and I'll do so again here (though with less gusto since I already lost the battle for the most part).  Minting a token is _not_ a transfer and therefore should _not_ fire a transfer event.  However, it _is_ interesting and an event __SHOULD__ be fired, but with a more appropriate name like `Mint`, `IncreaseSupply` or `Create`.  Similarly, when tokens are destroyed they are not transferred anywhere so firing a `Transfer` event is lying to listeners.  Instead it should fire a `Burn`, `DecreaseSupply` or `Destroy` event.

The reason I argue this is because a dapp author should not need to inspect the parameters of an event to figure out what the event _actually_ is.  e.g., is it a `Mint` event or a `Transfer` event.  This is akin to having a single event named `Event` whose first parameter is `eventType`.  We already have a mechanism for distinguishing events (the name of the event) and we don't need to overlay a second mechanism on top of it because we failed to leverage the name correctly.

The reason I lost this fight in the ERC20 spec is because by the time that spec actually got any activity, everyone had already gone off and implemented it (in various incarnations) and a handful of people implemented minting as `Transfer(from: 0x00..., ...)`.  The spec decided that since that strategy was already out in the wild, and it was too late to add a `Mint` and `Burn` event to the spec, it was preferable to codify that behavior rather than do nothing.  It is marked as __SHOULD__ because not all tokens do this.  Since it is __SHOULD__ instead of __MUST__, that means dapps cannot depend on it.  Since dapps cannot depend on it, I still argue that it doesn't add meaningful value to the spec.  Since it doesn't add value to the spec, I recommend token authors use an event name that is actually accurate so that dapps built targeting your token (where they _can_ depend on certain functionality) can behave reasonably and are easy to implement.
